pkg/sql/schemachanger/schemachanger_test: TestExecuteWithDMLInjection_alter_table_alter_column_type_validation failed

pkg/sql/schemachanger/schemachanger_test.TestExecuteWithDMLInjection_alter_table_alter_column_type_validation [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18361216?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/18361216?buildTab=artifacts#/) on master @ [75e6cfdb698031a6cfb2cc89044022979c65058f](https://github.com/cockroachdb/cockroach/commits/75e6cfdb698031a6cfb2cc89044022979c65058f):


```
=== RUN   TestExecuteWithDMLInjection_alter_table_alter_column_type_validation
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/a5015ccdc492420b5e82a2331f5ad6bc/logTestExecuteWithDMLInjection_alter_table_alter_column_type_validation1143295145
    test_log_scope.go:76: use -show-logs to present logs inline
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/28916/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/schemachanger/schemachanger_test_/schemachanger_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_type_validation/alter_table_alter_column_type_validation.definition:1:
        setup [0 args]
        CREATE TABLE t (i INT PRIMARY KEY, j TEXT);
        INSERT INTO t VALUES (1,NULL),(2,'FIT1'),(3,'FIT11');
        ----
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/28916/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/schemachanger/schemachanger_test_/schemachanger_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_type_validation/alter_table_alter_column_type_validation.definition:7:
        stage-exec [2 args]
        INSERT INTO t VALUES (10+$stageKey, 'too big for char(5)');
        ----
        pq: failed to satisfy CHECK constraint .*
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/28916/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/schemachanger/schemachanger_test_/schemachanger_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_type_validation/alter_table_alter_column_type_validation.definition:13:
        stage-exec [2 args]
        INSERT INTO t VALUES (20+$stageKey, 'rite');
        ----
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/28916/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/schemachanger/schemachanger_test_/schemachanger_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/schemachanger/testdata/end_to_end/alter_table_alter_column_type_validation/alter_table_alter_column_type_validation.definition:18:
        stage-query [2 args]
        SELECT count(*)=$successfulStageCount FROM t WHERE i > 3;
        ----
        true
    panic.go:626: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/a5015ccdc492420b5e82a2331f5ad6bc/logTestExecuteWithDMLInjection_alter_table_alter_column_type_validation1143295145
--- FAIL: TestExecuteWithDMLInjection_alter_table_alter_column_type_validation (7.10s)
=== RUN   TestExecuteWithDMLInjection_alter_table_alter_column_type_validation/injection_stage_{minOrdinal:1_maxOrdinal:1_phase:3_rollback:false}
    framework.go:482: query 'SELECT count(*)=$successfulStageCount FROM t WHERE i > 3' ($stageKey=(phase = PostCommitPhase stageOrdinal=1 rollback=false),$successfulStageCount=1): expected:
        true
        
        got:
        false
        
    cumulative.go:182: forcing job failure from BeforeStage due to test failure
    cumulative.go:177: short-circuiting BeforeStage hook due to test failure
    cumulative.go:237: 
        	Error Trace:	pkg/sql/schemachanger/sctest/cumulative.go:237
        	            				pkg/sql/schemachanger/sctest/test_server_factory.go:97
        	            				pkg/sql/schemachanger/sctest/cumulative.go:250
        	            				pkg/sql/schemachanger/sctest/cumulative.go:146
        	Error:      	Received unexpected error:
        	            	pq: transaction committed but schema change aborted with error: (XXUUU): t.Failed() is true
        	Test:       	TestExecuteWithDMLInjection_alter_table_alter_column_type_validation/injection_stage_{minOrdinal:1_maxOrdinal:1_phase:3_rollback:false}
    --- FAIL: TestExecuteWithDMLInjection_alter_table_alter_column_type_validation/injection_stage_{minOrdinal:1_maxOrdinal:1_phase:3_rollback:false} (2.84s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestExecuteWithDMLInjection_alter_table_alter_column_type_validation.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-45952

--- a/pkg/sql/schemachanger/sctest/cumulative.go
+++ b/pkg/sql/schemachanger/sctest/cumulative.go
@@ -193,6 +193,7 @@ func ExecuteWithDMLInjection(t *testing.T, relPath string, factory TestServerFac
 					defer jobErrorMutex.Unlock()
 					key := makeStageKey(s.Phase, s.Ordinal, p.InRollback || p.CurrentState.InRollback)
 					if _, ok := usedStages[key.AsInt()]; !ok {
+						t.Logf("Injecting into stage: %s", &key)
 						// Rollbacks don't count towards the successful count
 						if !p.InRollback && !p.CurrentState.InRollback &&
 							p.Stages[stageIdx].Phase != scop.PreCommitPhase {
@@ -243,6 +244,7 @@ func ExecuteWithDMLInjection(t *testing.T, relPath string, factory TestServerFac
 			// commits, this enforces any sanity checks one last time in
 			// the final descriptor state.
 			if lastRollbackStageKey != nil {
+				t.Logf("Job transaction committed. Re-inject statements from rollback: %s", lastRollbackStageKey.String())
 				injectionFunc(*lastRollbackStageKey, tdb, successfulStages)
 			}
 			require.Equal(t, errorDetected, schemaChangeErrorRegex != nil)

--- a/pkg/sql/schemachanger/sctest/framework.go
+++ b/pkg/sql/schemachanger/sctest/framework.go
@@ -443,6 +443,7 @@ func (e *stageExecStmt) Exec(
 		if len(stmt) == 0 {
 			continue
 		}
+		t.Logf("Starting execution of statment: %+v", stmt)
 		// Bind any variables for the statement.
 		boundSQL := os.Expand(stmt, func(s string) string {
 			switch s {

